File: src/main/java/com/yerti/core/inventories/InventoryHandler.java
package com.yerti.core.inventories;


import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;

/**
 * Listener used for IInventory menus
 */
public class InventoryHandler implements Listener {

    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (event.getInventory().getHolder() instanceof IInventory) {
            //Inventory is one of ours
            event.setCancelled(true);
            IInventory inventory = (IInventory) event.getInventory().getHolder();
            inventory.onGUI((Player) event.getWhoClicked(), event.getRawSlot(), event.getCurrentItem(), event);
        }

        if (event.getWhoClicked().getOpenInventory().getTopInventory() != null && event.getWhoClicked().getOpenInventory().getTopInventory().getHolder() instanceof CustomInventory) {
            if (event.isShiftClick()) {
                event.setCancelled(true);
            }
        }

        if (event.getInventory().getHolder() instanceof CustomInventory) {
            CustomInventory inventory = (CustomInventory) event.getInventory().getHolder();
            if (inventory.cancelsEvent() && event.getRawSlot() < inventory.getSlots()) event.setCancelled(true);
            System.out.println(inventory.getItems().size());
            if (inventory.getItems().containsKey(event.getSlot())) {
                inventory.getItems().get(event.getSlot()).onClick(event);
            }

        }
    }


}

File: src/main/java/com/yerti/core/inventories/InventoryHandler.kt
package com.yerti.core.inventories

import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent

/**
 * Listener used for IInventory menus
 */
class InventoryHandler : Listener {
    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        if (event.inventory.holder is IInventory) { //Inventory is one of ours
            event.isCancelled = true
            val inventory = event.inventory.holder as IInventory
            inventory.onGUI(event.whoClicked as Player, event.rawSlot, event.currentItem, event)
        }
        if (event.whoClicked.openInventory.topInventory != null && event.whoClicked.openInventory.topInventory.holder is CustomInventory) {
            if (event.isShiftClick) {
                event.isCancelled = true
            }
        }
        if (event.inventory.holder is CustomInventory) {
            val inventory = event.inventory.holder as CustomInventory
            if (inventory.cancelsEvent() && event.rawSlot < inventory.slots) event.isCancelled = true
            println(inventory.getItems().size)
            if (inventory.getItems().containsKey(event.slot)) {
                inventory.getItems().getValue(event.slot).onClick(event)
            }
        }
    }
}